apps/design/shared/src/layout.ts

```diff
@@ -1641,3 +1706,74 @@ export function layOutBallot(
     return wrapException(e);
   }
 }
+
+interface LayoutAllBallotsParams {
+  election: Election;
+  isTestMode: boolean;
+}
+
+function layOutAllBallotsHelper({
+  election,
+  isTestMode,
+  electionHash,
+}: LayoutAllBallotsParams & { electionHash?: string }): BallotLayout[] {
+  return election.ballotStyles.flatMap((ballotStyle) =>
+    ballotStyle.precincts.map((precinctId) => {
+      const precinct = assertDefined(getPrecinctById({ election, precinctId }));
+      return layOutBallotHelper({
+        election,
+        precinct,
+        ballotStyle,
+        isTestMode,
+        electionHash,
+      });
+    })
+  );
+}
+
+/**
+ * Given an election without gridLayouts, lays out all of the ballots for the
+ * election (for every ballot style/precinct combo). Returns the laid out
+ * ballots as well as the election definition with gridLayouts.
+ *
+ * We have to lay out all of the ballots at once in order to make sure the
+ * correct election hash is encoded in the ballot metadata. The election hash
+ * needs to be from the completed election with gridLayouts defined.
```

Comment:
> We have to lay out all of the ballots at once in order to make sure the correct election hash is encoded in the ballot metadata.

I initially read this as meaning that we're performing a check on election hash in the first pass, which isn't the case. We're just preparing the data that we need to compute a proper election hash. I'm parsing the sentence correctly now, after reading the third paragraph in this comment.

I think you could even drop this paragraph altogether. The third paragraph was the one that gave me the most clarity!

Reply:
Great, will change